File: 백준/Gold/14725. 개미굴/개미굴.cc
#include <iostream>
#include <string>
#include <vector>
#include <map>
using namespace std;

int N, K, T;
class Node {
public:
	string val;
	map<string, Node*> child;

	Node() {}
	Node(string val)
	{
		this->val = val;
	}
};

class Trie {
public:
	Node* Root;

	Trie()
	{
		this->Root = new Node("Root");
	}

	void insert(Node* pos, vector<string>& inputVec, int idx)
	{
		if (idx >= (int)inputVec.size())
		{
			return;
		}

		string val = inputVec[idx];
		if (pos->child.count(val) == 0) // 없을 경우 새로 노드 생성하여 자식에 추가
		{
			pos->child.insert({ val, new Node(val) });
		}

		/* 재귀로 진행 */
		insert(pos->child[val], inputVec, idx + 1);
	}

	void print(Node* pos, int depth)
	{
		if (pos->val != "Root")
		{
			for (int i = 1; i < depth; i++)
			{
				cout << "--";
			}
			cout << pos->val << '\n';
		}

		for (auto iter = pos->child.begin(); iter != pos->child.end(); iter++)
		{
			print(iter->second, depth + 1);
		}
	}
};

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	cin >> N;

	Trie trie;
	string str;
	for (int n = 0; n < N; n++)
	{
		cin >> K;
		vector<string> inputVec;
		for (int k = 0; k < K; k++)
		{
			cin >> str;
			inputVec.push_back(str);
		}
		trie.insert(trie.Root, inputVec, 0);
	}

	trie.print(trie.Root, 0);
	return 0;
}

File: 백준/Gold/14725. 개미굴/개미굴.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        Trie trie = new Trie("ROOT");
        int N = Integer.parseInt(st.nextToken());

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int K = Integer.parseInt(st.nextToken());

            List<String> inputList = new ArrayList<>();
            for (int k = 0; k < K; k++) {
                inputList.add(st.nextToken());
            }
            trie.insert(trie.root, inputList, 0);
        }
        trie.print(trie.root, -1);
    }

    public static class Trie {
        Node root;

        public Trie() {
            this.root = null;
        }

        public Trie(String val) {
            this.root = new Node(val);
        }

        public void insert(Node pos, List<String> inputList, int idx) {
            if (idx >= inputList.size()) {
                return;
            }

            String val =  inputList.get(idx);
            if (pos.child.get(val) == null) { // 없을 경우 새로 생성하여 자식에 추가
                pos.child.put(val, new Node(val));
            }
            insert(pos.child.get(val), inputList, idx + 1);
        }

        public void print(Node pos, int depth) {
            if (pos.val != "ROOT") {
                for (int i = 0; i < depth; i++) {
                    System.out.print("--");
                }
                System.out.println(pos.val);
            }
            for (Node child : pos.child.values()) {
                print(child, depth + 1);
            }
        }
    }

    public static class Node {
        String val;
        Map<String, Node> child;

        public Node(String val) {
            this.val = val;
            this.child = new TreeMap<>();
        }
    }
}